### ISSUE TITLE
Netflix not working in nwjs, Widevine component

### ISSUE THREAD

**Issue by krneticz** [nwjs/nw.js#6425]
Currently, Netflix does not work in nwjs due to the missing Widevine component. Is there a way for it to be added as the plugin or otherwise?

**Comment by rogerwang** [comment]
Yeah it should be enabled. Will look into it. Is there anyway to test it without any account? The service is not available here.

**Comment by krneticz** [comment]
I tried to lookup way to test it without Netflix account, but no luck. Currently, our users in the UK when tried to play Netflix in our built-in browser (webview) are getting the WideVine missing message. That is in nwjs 27 in webview. If you have some recommendations how to try to make it work, I can try and test with Netflix.

**Comment by halton** [comment]
@rogerwang I think the Widewine support only available on Google Chrome browser. Did NW.js add hack to enable it? Like https://bugs.launchpad.net/ubuntu/+source/chromium-browser/+bug/1371274.

**Comment by alexgor1** [comment]
@rogerwang you can test it with DRM video by this link https://demo.castlabs.com/

**Comment by GnorTech** [comment]
A test build with this feature will be available soon.

**Comment by GnorTech** [comment]
Please try this build: https://dl.nwjs.io/live-build/nw29/02-10-2018/6aee6eb-43410a6-c0aa57f-1b9d678/v0.29.0-beta1/
It should work if you put the plugin ('WidevineCdm' directory of Chrome) in the directory of NW. I didn't test it in webview though.

**Comment by krneticz** [comment]
I've tried but Netflix is still complaining about missing Widevine. I've added folder from Chrome 64, maybe that is a problem? Should it be originating from the specific version of Chrome? Is there something to be added as chromium arguments, like we are adding for the Flash plugin?

**Comment by GnorTech** [comment]
I used the files from Chrome 64 too. Note that the architecture of the binary file should match: NW.js win-x64 with the win-x64 plugin binary. After that you could see the plugin is listed in `navigator.plugins`. And `nw.exe --url=https://demo.castlabs.com/` works for me. The video is playable.

**Comment by krneticz** [comment]
Yes, I added Widevine 64bit to 64bit SDK nwjs. That demo (https://demo.castlabs.com/) works for me too- but it also works without the Widevine folder, so I think that is not a relevant test case. BTW, when I open that page, both with and without the Widevine folder, I see "Selected DRM: none" in the console logs.

**Comment by alexgor1** [comment]
Netflix works ok on Mac OS and Linux for this build, but it has new problem with webview on Linux which was absent: 
After first resizing of webview it stops to render. So i can load new URL to it and even debug it, but webview's layout shows rendered frame from previous page until i resize webview again. After resizing it refreshes rendered frame, but still doesn't response to any user actions and doesn't render any changes untill next resizing.

**Comment by GnorTech** [comment]
Is there any test case other than Netflix?

**Comment by krneticz** [comment]
I have just tested nw 0.29.0 win64 sdk on Windows 8.1 64 bit with WidevineCdm from Chrome 65 64bit placed in nwjs folder, and Netflix is still complaining about missing WidevineCdm.

<img width="1057" alt="screen shot 2018-03-09 at 12 39 39 pm" src="https://user-images.githubusercontent.com/281657/37205981-1bdeeada-2397-11e8-9a6f-1e22a4ff4afe.png">

**Comment by rogerwang** [comment]
Please provide a case other than Netflix. We cannot use it here. Without a proper case we're not able to move forward...

**Comment by krneticz** [comment]
If test case is needed, Netflix offers free 30 days trial, and if outside of US https://www.smartdnsproxy.com can help and also offers the trial.

**Comment by rogerwang** [comment]
@krneticz it's really not productive to work around Netflix's location detection... AFAIK they worked hard to ensure it.

**Comment by krneticz** [comment]
@rogerwang I understand, it does require extra effort, I'm doing it since I'm located in Serbia (smartdnsproxy works for me). Unfortunately, I just cannot find any other test case!

**Comment by Virius** [comment]
Hi. On nw.js 0.29.9 and 0.30.0 (sdk,  windows 10 x64) demo (https://demo.castlabs.com/) not work

**Comment by krneticz** [comment]
SeeYouLink team just placed the bounty on this issue fix https://www.bountysource.com/issues/54253661-netflix-not-working-in-nwjs-widevine-component

cc @rogerwang @GnorTech

**Comment by krneticz** [comment]
@rogerwang It's been a long time, but it seems that we have Netflix demo case that is accessible from anywhere in the world without Netflix account: https://www.netflix.com/title/80018499

I just tested from Serbia, no account on Windows 8 machine. Tests work in Chrome browser but fail in SeeYouLink app based on nwjs 0.40.2

**Comment by halton** [comment]
I think Netflix not only check the UA, but also the API, will lead our app based on Chromium63 is not working as well. So upgrade chromium core might be helpful.

**Comment by rogerwang** [comment]
Thanks for the link @krneticz , will look into it.

btw, does it work in the Chromium open source browser (not Chrome)?

**Comment by krneticz** [comment]
Unfortunately, it does not :( I tried the latest Chromium on Windows 10, and it fails to play the test. I looked bit about the issue, and it seems that widevine component (dll?) from Chrome can be added to Chromium, but don't have exact recipe or clear case.

**Comment by rogerwang** [comment]
@halton I'm confused by your comment, as your company's app is not based on NW as far as I know.

**Comment by halton** [comment]
@rogerwang yes, you're right. But we're dealing with similar issue on CDM, that's why I'm here.

**Comment by panther7** [comment]
Widevine doesn't work on NWjs based app too.

Latest NWjs with included Widevine binary: https://drive.google.com/open?id=1Nd3N0-m3GtDCieXJVfO_R8-huLz0qcu3

**Comment by panther7** [comment]
> 
> 
> @halton I'm confused by your comment, as your company's app is not based on NW as far as I know.

Widevine doesn't work on NWjs based app too.

Latest NWjs with included Widevine binary: https://drive.google.com/open?id=1Nd3N0-m3GtDCieXJVfO_R8-huLz0qcu3

**Comment by matthiasfostel** [comment]
The Castlabs demo has two Widevine DRM video examples listed, scroll down for the two Big Bucky videos. They have the DRM icon next to them. All examples are H.264 encoded though.

https://demo.castlabs.com/

Chromium can be compiled with Widevine support. Versions built with support can be found here:

https://chromium.woolyss.com/

**Comment by anesuc** [comment]
ny update on this?

**Comment by Virius** [comment]
Hi! Will there be any changes on this issue? DRM worked on NW.js v0.35.4, but stopped working on next versions. Checked on v0.44.6.

Steps for testing:
1. replace the ffmpeg library to version with proprietary codecs
2. copy WidevineCdm folder from chrome
3. run: nw.exe --url=https://demo.castlabs.com/
4. Next, select the video marked DRM on the page.

maybe there are some workarounds?

**Comment by nitrram** [comment]
Hi there,

I've been too lazy to explore and to get know all the flags regarding WidevineCdm, but would give it a try and would have been successful -> up and running on Win/OSX/GNU\/Linux.

So I build NW.js patched similarly to the patch of nw46 branch I attach here.

It can be used as a hint at least...

tl;dr;
I've got rid off preprocessor directives related to Widevine and its components. Plus I perhaps had to paste correct version of the WidevineCdm bundle in the
`third_party/widevine/cdm/widevine_cdm_version.h`

Then I added `WindevineCdm/` dir up to the nw executable level...

[widevine.patch.gz](https://github.com/nwjs/nw.js/files/4580994/widevine.patch.gz)

**Comment by krneticz** [comment]
@rogerwang is there a chance that somebody can try to incorporate patch by @nitrram into nwjs codebase any time soon?

**Comment by rogerwang** [comment]
Sure, will look into it.

**Comment by krneticz** [comment]
@rogerwang any luck with patch by @nitrram ?

**Comment by krneticz** [comment]
@rogerwang is there a chance that somebody can try to incorporate patch by @nitrram into nwjs codebase any time soon?

**Comment by nitrram** [comment]
I'll take a look at the stuff more thoroughly this weekend and will try to emit a pull/merge request for this. I will report then

**Comment by krneticz** [comment]
@nitrram thank you very much!

**Comment by nitrram** [comment]
I have pushed a proposal solution into a fork of nwjs/chromium.src, but yet I'm bit struggling with BUILDFLAG macros.

The proposal is getting rid of the widevine bundle/component/self update machinery and its idea is to simplify chromium following way:
1) if widevine_enabled defined and true within args.gn, incorporate all of the code regarding widevine
2) disable widevine component updating mechanism
3) if widevine_enabled, chromium is widevine ready but in order to get it work you should add `WidevineCdm` directory with its content to the place the platform expects (on Linux and Windows it's on the binary's level, on MacOS I don't remember exactly).

It's far from ideal solution, but you don't have to change WidevineCdm so often, and I'm not sure of the need to have exactly same version in the widevine version header and its binaries (I have a feeling, that some versions were compatible enough overcoming the version mismatch). But in such a case, you should check the version of widevine a specific release of nw was built with and supply it with it (e.g. taken from google-chrome).

If it is an utter nonsense or garbage at first sight :), let me know. I would investigate it further on, but I'd need some concrete proposal of how it should be done in such a case. 

If it's ok, I'll fix the macros and emit a pull request on that.

**Comment by matthiasfostel** [comment]
Seems like a good start but adding the ability for it to check the App.dataPath for the WidevineCmd folder would make it a lot easier to update for developers.

**Comment by anesuc** [comment]
Any updates on this yet?

**Comment by Virius** [comment]
has any updates? This problem is real blocker for us which prevents us from updating to the last nw.js

**Comment by nitrram** [comment]
Sorry guys being negative due to several reasons. I have no time to spare since my family has gone bigger recently. The other and also the more significant one is, that Widevine is licensed - it should be worked out individually (per a fork of Chromium/nwjs). Well, I hate DRM itself, but the thing is that whatever hack I can do, is not sustainable. Just a few weeks ago I ran into some problems to make Widevine up in the recent version of Electron. There is a slightly different approach of registering content but I was not successful doing the same thing in there. Correct me if I'm wrong...

**Comment by anesuc** [comment]
@nitrram Addressing the licensing issue, I think a solution to this would be to make the individuals download the licensed components separately. Equivalent to how codecs are handled. That way, the burden of handling the licensing side of things is on the user/developer. Not sure how difficult that would be to do things that way though

**Comment by rogerwang** [comment]
Supported in this build: https://dl.nwjs.io/live-build/nw50/20201122-032800/431d6295f/v0.50.1/ . Two new APIs are provided that the plugin can be downloaded on end-users' side.

sample code:
```javascript
function openNewWin() {
   nw.Window.open('https://shaka-player-demo.appspot.com/demo/');
}

nw.App.enableComponent('WIDEVINE', function(version) {
   console.log('version: ' + version);
   if (version === '0.0.0.0') {
     nw.App.updateComponent('WIDEVINE', function(success) {
        openNewWin();
     });
   } else {
     openNewWin();
   }
});
```
Note that the new APIs are experimental and subject to change.

**Comment by anesuc** [comment]
Ok, this is great news! Is the widevine plugin Chrome version dependant?

**Comment by rogerwang** [comment]
It depends a little bit. You can check the dependency in the manifest file of the plugin, which resides in the user data directory.

**Comment by SaulGDev** [comment]
> Supported in this build: https://dl.nwjs.io/live-build/nw50/20201122-032800/431d6295f/v0.50.1/ . Two new APIs are provided that the plugin can be downloaded on end-users' side.
> 
> sample code:
> 
> ```js
> function openNewWin() {
>    nw.Window.open('https://shaka-player-demo.appspot.com/demo/');
> }
> 
> nw.App.enableComponent('WIDEVINE', function(version) {
>    console.log('version: ' + version);
>    if (version === '0.0.0.0') {
>      nw.App.updateComponent('WIDEVINE', function(success) {
>         openNewWin();
>      });
>    } else {
>      openNewWin();
>    }
> });
> ```
> 
> Note that the new APIs are experimental and subject to change.

So Widevine component from the same version of Chrome won't work, as Widevine will fail license check? That's what's happening for me on Netflix/Spotify and on the other hand, the test demo https://bitmovin.com/demos/drm  plays just fine.

**Comment by rogerwang** [comment]
Netflix should work as long as you have the propietary codecs: https://nwjs.readthedocs.io/en/latest/For%20Developers/Enable%20Proprietary%20Codecs/

**Comment by panther7** [comment]
> Supported in this build: https://dl.nwjs.io/live-build/nw50/20201122-032800/431d6295f/v0.50.1/ . Two new APIs are provided that the plugin can be downloaded on end-users' side.
> 
> sample code:
> 
> ```js
> function openNewWin() {
>    nw.Window.open('https://shaka-player-demo.appspot.com/demo/');
> }
> 
> nw.App.enableComponent('WIDEVINE', function(version) {
>    console.log('version: ' + version);
>    if (version === '0.0.0.0') {
>      nw.App.updateComponent('WIDEVINE', function(success) {
>         openNewWin();
>      });
>    } else {
>      openNewWin();
>    }
> });
> ```
> 
> Note that the new APIs are experimental and subject to change.

Does not work in NWjs 0.52... 

```js
nw.App.enableComponent('WIDEVINE', console.log); // '0.0.0.0'
nw.App.updateComponent('WIDEVINE', console.log); // false
```
